src/Authentication/Aggregate/User.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+namespace Authentication\Aggregate;
+
+use Authentication\ReadModel\EmailIsRegistered;
+use Authentication\ReadModel\IsUserBlocked;
+use Authentication\Service\NotifyOfIntrusionDetection;
+use Authentication\Value\ClearTextPassword;
+use Authentication\Value\EmailAddress;
+use Authentication\Value\PasswordHash;
+use Webmozart\Assert\Assert;
+
+class User
+{
+    /** @var EmailAddress */
+    private $email;
+
+    /** @var PasswordHash */
+    private $passwordHash;
+
+    private function __construct(EmailAddress $email, ClearTextPassword $password)
+    {
+        $this->email        = $email;
+        $this->passwordHash = $password->makeHash();
+    }
+
+    public static function register(
+        EmailAddress $email,
+        ClearTextPassword $password,
+        EmailIsRegistered $emailIsRegistered
+    ) : self {
+        Assert::false($emailIsRegistered->__invoke($email));
+
+        return new self($email, $password);
+    }
+
+    public function authenticate(
+        ClearTextPassword $password,
+        IsUserBlocked $blockedUsers,
+        NotifyOfIntrusionDetection $intrusionDetection
+    ) : void {
+        Assert::true($password->verify($this->passwordHash));
+
+        if ($blockedUsers->__invoke($this->email)) {
```

Comment:
Can I ask.. Why magic method `__invoke` is used, but it is not called like this `$blockedUsers($this->email)`?

Reply:
Good question: this is because of an IDE issue. This is a known issue of PHPStorm, which cannot detect usages of callable objects used with the syntax that you proposed.

While it is true that I shouldn't please my tools too much, the IDE is such a massive central point in my day-to-day operations (especially around refactoring/inspection) that I'd rather keep the `__invoke` explicit until the bug is solved (should already be for PHPStorm 2019.02)

Reply:
Wow such a quick reply :) Ok I understand. Thank you!